Add array as an alternative input + `code` option

gforge opened the pull request:
This PR does the following:

- Allows for `keyValue`/`keyCode` to come in array format
- Changes behavior to allow simultaneous input of all three options (**API change**) see discussion below &dagger;
- Adds the option of having `code` instead of `keyCode` (also marks `keyCode` as deprecated)
- Adds an example for toying around with `code`
- Clarifies the examples by visually boxing each example and adding code sections where appropriate
- Refactoring of the code so that `index.js` is more informative, decorators have a separate file + more
- Updated the README and also entered info on the missing section (much is copy-paste from MDN)

The PR includes all the updates in previous Flow PR (https://github.com/ayrton/react-key-handler/pull/147)

&dagger; The possibility of `keyValue` having priority is problematic in my opinion. If a user does not intend to use both then they probably wouldn't have specified them. If this truly should not be allowed then the constructor should throw an error. I can't see any use case where allowing all provided value hooks to trigger the event function.

gforge committed:
Updated packages and fixed unit tests

gforge committed:
Eslint fixes

gforge committed:
Fixed all flow errors

gforge committed:
Updated webpack production build

gforge committed:
Fixed webpack loaders

gforge committed:
Added flow interface files (issue #26)

gforge committed:
WIP: add code + array format

restructured parts of the code for better overview

gforge committed:
Fixed the array + code option

gforge committed:
Updated travis node environment

gforge committed:
Trying to fix the outdated yarn version bug

gforge committed:
Merge branch 'fix_flow' into add_code_and_array

gforge committed:
Updated README

gforge committed:
Fixed readme links and prop description

gforge committed:
Improved the demos

gforge committed:
Removed the gforge specifics prior to PR

gforge committed:
Renamed other to otherProps

gforge committed:
Isolated the window.KeyboardEvent() call

gforge committed:
Resolve merge with fix_flow branch

gforge committed:
Fixed and improved demos

Added syntax highlighter
Added h3 style for Code:
Fixed wrong code example
Code-explorer no always has 3 elements

gforge committed:
Removed yarn update

gforge committed:
Merge branch 'fix_flow' into add_code_and_array

gforge commented on demo/components/examples/decorators/key-handler.js:
```diff
@@ -22,8 +23,15 @@ function Demo({keyValue}: Props) {
 }
 
-export default keyHandler({keyEventName: KEYPRESS, keyValue: 's'})(Demo);
+export default keyToggleHandler({keyEventName: KEYPRESS, keyValue: 's'})(Demo);
```
Fixed in new example. I've also added the `react-syntax-highlighter` package and an h3-style to make the examples easier to read.

gforge commented:
I would love your opinion on the file naming and structure. There are a few things that I've been thinking about:

- `key-handler-decorator`: Should probably just be `decorator`
- `utils.js`: This file is stuffed with functions. I personally prefer having a folder with an index file that merges all the functions.
- `keyHandler` and `keyToggleHandler`: These should perhaps be defined in the `index.js`

I've been using `rollup` for some of my libraries and it allows more `mjs` and other output formats that could be interesting optimizations: https://medium.com/@tarkus/how-to-build-and-publish-es6-modules-today-with-babel-and-rollup-4426d9c7ca71

gforge committed:
Merged upstream

gforge committed:
Forgot about flow among the constants

gforge committed:
Added Flow to constants

gforge committed:
The Flow restriction on the onKeyHandler was too strict

ayrton commented on README.md:
```diff
@@ -2,7 +2,7 @@
 
```
These README changes are +💯

ayrton commented on lib/index.js:
```diff
@@ -1 +1,6 @@
+export {KEYDOWN, KEYPRESS, KEYUP} from './constants';
+export {default as keyHandleDecorator, keyToggleHandler, keyHandler} from './key-handle-decorator';
+
+export default KeyHandler;
```
Won't the end-user now have to use something like webpack? (vs. when using `module.exports`)

gforge replied:
I don't think so but if we go with rollup it's easy to deliver umd, es6, cjs and other packages

ayrton replied:
Rollup definitely wasn't as popular as it is now when this package was initially released, seems most libraries moved to it, and sounds like we should too

gforge replied:
Sure, I will prepare a PR for it once we're done with this. The docs make it sound like it should be super easy but then you start looking into all the options and its a :rabbit2: :hole: I have a good setup in a similar project although in that one I used Typescript so it needs some tweaking before we can deploy it.

I'm not a great fan but I think webpack-dev-server is super convenient for the demo and therefore we may want to keep it in place.

ayrton commented on lib/key-handle-decorator.js:
```diff
@@ -0,0 +1,64 @@
+import { type rawOrArray } from './types';
+
+export type KeyhandleDecoratorState = {|
+ keyValue?: ?rawOrArray<string>,
```
Wonder if there's a better name for `raw`, or is this something commonly used?

ayrton replied:
I like `primitiveOrArary`

gforge replied:
Fixed

ayrton commented on lib/key-handle-decorator.js:
```diff
@@ -0,0 +1,64 @@
+
+        render() {
+          return (
+            <div>
```
We should probably do `React.Fragment` here, wdyt?

ayrton commented:
@gforge 

> I would love your opinion on the file naming and structure. There are a few things that I've been thinking about:

Let's discuss this in an issue or separate PR

gforge commented on lib/utils.js:
```diff
@@ -100,15 +35,33 @@ function isContentEditable(element: HTMLElement): boolean {
  */
 
-export function matchesKeyboardEvent(event: KeyboardEvent, {keyCode, keyValue}: KeyboardKey): boolean {
+function elmntOrArrayMatch<T>(a: T | T[], b: T) {
```
This is just a helper to the original matchesKeyboardEvent, not sure I understand

ayrton replied:
Suggesting a function rename, to make it more in lines of the "convention" of the other function `matchesKeyBoardEvent`

gforge replied:
Excellent suggestion, changed

gforge commented on package.json:
```diff
@@ -68,6 +68,7 @@
     "exenv": "^1.2.0",
-    "prop-types": "^15.6.2"
+    "prop-types": "^15.6.2",
+    "react-syntax-highlighter": "^8.0.1"
```
Will change, writing on the phone.

gforge commented on test/support/helper.js:
```diff
@@ -3,8 +3,6 @@ import chaiEnzyme from 'chai-enzyme';
 import Adapter from 'enzyme-adapter-react-16';
 import { JSDOM } from 'jsdom';
-import path from 'path';
-import requirePaths from 'app-module-path';
```
The problem was that my IDE complained that it couldn't find the utils package.

I think though that tests should be super easy to understand. The app-module-path requires that you first check the setup and also understand the package. 

As a side note, I wasn't sure that I would get this great interaction. Much of the changes were made to adapt to my own coding style as I was certain to post my own scoped version of the package. There is an abundance of abandoned packages out there, given how annoying updating all the dependencies are, it's quite understandable :grin:

ayrton replied:
Let's remove the `app-module-path` dependency

gforge replied:
Fixed

gforge committed:
PR fixes

Renamed rawOrArray
Fixed React.Fragment instead of div
Corrected syntax highlighter to dev deps

gforge commented:
Updated the PR with the fixes

ayrton commented on lib/key-handle-decorator.js:
```diff
@@ -0,0 +1,64 @@
+
+        render() {
+          return (
+            <React.Fragment>
```
We'll have to update our peerDependencies in package.json:

``` json
{
  "peerDependencies": {
    "react": ^16.2.0-0"
  }
}
```

gforge replied:
Fixed

gforge committed:
Renamed to matchesElementOrArray, removed dep & updated React dependency

ayrton committed:
Set up rollup

gforge committed:
Merge with parcel-rollup

ayrton commented:
@gforge can you rebase, would love to get this in for 1.2.0

gforge committed:
Resolved merge conflicts

gforge commented:
Ok, rebased from master

ayrton commented on README.md:
```diff
@@ -124,32 +128,43 @@ The prop types of the `KeyHandler` component are:
 
 ## Key event names
 
-TODO: explain the differences between the different key events.
```
Can you leave the above and add a new section for the new documentation you've added: 

```md
## `keyValue'`, `code` and `keyCode`
```

We'll also need to update the TOC because the heading changed

gforge committed:
Updated README contents + added a code type specific header for README

gforge committed:
Added automated TOC

gforge committed:
Removed the Testimoneals and TOC from TOC

gforge committed:
Minor README fix

ayrton commented on README.md:
```diff
@@ -124,32 +131,47 @@ The prop types of the `KeyHandler` component are:
+
+The three available key events are
+
+- `key`      This corresponds to the true value. This is the value of the key pressed
```
this should be `keyValue` vs. `key`

ayrton commented on lib/key-handler.js:
```diff
@@ -34,8 +46,11 @@ export default class KeyHandler extends React.Component<Props> {
 
-    if (!props.keyValue && !props.keyCode) {
+    if (!props.keyValue && !props.keyCode && !props.code) {
       console.error('Warning: Failed propType: Missing prop `keyValue` or `keyCode` for `KeyHandler`.');
```
Can you update the error message to include `code`

ayrton commented on lib/key-handler.js:
```diff
@@ -34,8 +46,11 @@ export default class KeyHandler extends React.Component<Props> {
 
     /* eslint-disable no-console */
+    if (props.keyCode) {
+      console.warn('The `keyCode` is deprecated, use `code` instead');
```
```js
console.warn('Warning: Deprecated propType: `keyCode` is deprecated in favour of `code` for `KeyHandler`.');
```

ayrton commented on lib/key-handle-decorator.js:
```diff
@@ -0,0 +1,64 @@
+                keyEventName={keyEventName}
+                onKeyHandle={this.handleKey}
+              />
+              <Component {...otherProps} {...this.state} />
```
this should be `<Component {...this.props} {...this.state} />` otherwise the decorated component will not receive their original props

ayrton commented on lib/key-handler.js:
```diff
@@ -4,20 +4,32 @@ import React from 'react';
-  +keyCode?: number,
-  +keyEventName: KEYDOWN | KEYPRESS | KEYUP,
-  +onKeyHandle?: (event: KeyboardEvent) => void,
+  +keyValue?: ?primitiveOrArray<string>,
```
What are the `+` prefixes for?

ayrton replied:
TIL they indicate the props are readonly, see flow docs https://flow.org/en/docs/types/interfaces/#toc-interface-property-variance-read-only-and-write-only

gforge replied:
Sorry, wrote the answers yesterday but there was a "pending" set to the answer. TIL - you need to submit pending comments at the top of the page :grimacing:

ayrton commented on test/components/key-handler.test.js:
```diff
@@ -71,7 +82,10 @@ describe('KeyHandler', () => {
+  // This make little sense - a user will not understand that there is a priority
+  // the user wil most likely expect that the function gets called if any
+  // of the matching criteria match.
+  it.skip('prioritizes key value over code', () => {
```
If this makes no sense we should `invariant` in the constructor

ayrton replied:
It's an [npm package](https://www.npmjs.com/package/invariant)

ayrton commented on test/components/key-handler.test.js:
```diff
@@ -19,13 +19,24 @@ describe('KeyHandler', () => {
+    mount(<KeyHandler keyValue={'s'} onKeyHandle={handler} />);
 
-    triggerKeyEvent(KEYUP, undefined, ARROW_LEFT);
+    triggerKeyEvent(KEYUP, undefined, 's');
```
why were these changed?

gforge committed:
Fixed minor issues before merge

gforge committed:
Package update to newest flow version

gforge committed:
Upgraded to babel 7

ayrton commented:
Why the 5bcdd9a change? Since it's no official release I'd just revert that commit

ayrton commented on test/components/key-handler.test.js:
```diff
@@ -71,7 +82,10 @@ describe('KeyHandler', () => {
+  // This make little sense - a user will not understand that there is a priority
+  // the user wil most likely expect that the function gets called if any
+  // of the matching criteria match.
+  it.skip('prioritizes key value over code', () => {
```
yep I'm okay with this braking change, dropping the test makes sense once we throw the invariant

gforge commented:
Thanks for merging! Sorry, didn't check babylon & parcel compatibility. Also, I like the prettier code style change :smile:

ayrton commented:
@gforge no worries, you did an epic job here! Just released v1.2.0-beta.3 so feel free to test it out

gforge commented:
Thanks. I've tested it and it seems to work fine.

I think you should do a major release since the handling of simultaneous keycodes has changed
